Technical / editorial: Document the text-box-* properties (#37738)
mdn/content: files/en-us/web/css/text-edge/index.md

---
title: <text-edge>
slug: Web/CSS/text-edge
page-type: css-type
browser-compat: css.types.text-edge
---

{{CSSRef}}

The **`<text-edge>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/CSS/CSS_Types) specifies font metrics representing specific regions on a font's start block edge and end block edge that are to be manipulated in some way. It effectively does this by specifying a position for the font's over and/or under edge.

> `text-edge` values are used in various CSS properties. For example, in the {{cssxref("text-box-edge")}} property it is used to define regions of the font {{glossary("leading")}} to trim.

Author: I've used most of your suggestions here, but I've kept it as two paragraphs; I think it is easier to read that way.

The `<text-edge>` data type is composed of one or two keywords representing specific regions on a font's start block edge and end block edge. The available keywords are as follows:

- `alphabetic`
  - : The font's under edge is its alphabetic baseline, which is the bottom of its short lower-case letters (for example, "m", "n", and "o") or capital letters.
- `cap`
  - : The font's over edge is its cap-height baseline, which is the top of its capital letters.
- `ex`
  - : The font's over edge is its x-height baseline, which is the top of its short lower-case letters.
- `text`
  - : The font's over and under edges are its start and end boundaries. This can be considered the default, as this is where the boundaries are normally without any manipulation.

If two values are specified, the first value specifies the position of the font's over edge, and the second value specifies the position of the font's under edge.

- Valid over edge values: `text` (the default), `cap`, and `ex`.
- Valid under edge values: `text` (the default) and `alphabetic`.

If one value is specified, it specifies the over _and_ under edge position. At the time of writing, the only valid single value is `text`.

> 

Author: Right, I've updated it to use something more like the structure you've specified, but:

- I've used "single keyword values" and "two keyword values". I found "Double edge" and "Secondary edge" really confusing.
- I've kept my note about the unsupported values. I don't see the point in providing definitions when their behavior is currently not agreed upon in the WG.

> The specification also defines the `ideographic` and `ideographic-ink` keywords, which are intended to represent over and under edge positions specific to [CJK language characters](https://en.wikipedia.org/wiki/CJK_characters). Currently their exact behavior is being debated and they are are not supported by any browser.

Author: I've kept this, as per my above statement.

## Examples

### Valid text-edge values

```css
/* Single keyword */
text-box-edge: text;

/* Two keywords */
text-box-edge: text text;
text-box-edge: text alphabetic;
text-box-edge: cap alphabetic;
text-box-edge: ex text;

> ```

Author: updated

- When two values are specified, the first value specifies the position of the font's over edge, and the second value specifies the position of the font's under edge.

### Single keyword values

- `text`
  - : The font's over and under edges are its start and end boundaries. This can be considered the default, as this is where the boundaries are normally without any manipulation.

> [!NOTE]
> The specification also defines the `ideographic` and `ideographic-ink` keywords, which are intended to specify over and under edge positions specific to [CJK language characters](https://en.wikipedia.org/wiki/CJK_characters). Currently their exact behavior is being debated and they are are not supported by any browser.

### Two keyword values

- `alphabetic`
  - : The font's under edge is its alphabetic baseline, which is the bottom of its short lower-case letters (for example, "m", "n", and "o") or capital letters.
- `cap`
  - : The font's over edge is its cap-height baseline, which is the top of its capital letters.
- `ex`
  - : The font's over edge is its x-height baseline, which is the top of its short lower-case letters.
- `text`

>   - : The font's over or under edge is its start and end boundaries. This can be considered the default, as this is where the boundaries are normally without any manipulation.

Reviewer: This is a tough one, "start and end boundaries" look a bit ambiguous to me. I wonder if there are better ways to describe this. How about adding "for text runs, this means font's ascent and descent, without half-leading"? Please disregard if this could make it even more confusing...

Author: I think this is a useful detail to include, as you are right, the existing definition is a bit ambiguous/vague. I've changed the first description of `text` in the "Single keyword values" section to:

> The font's over and under edges are its start and end boundaries, which include the font's ascenders and descenders, but exclude the half-leading set on the text. This can be considered the default, as this is where the boundaries are normally without any manipulation.

I've then changed the second definition of `text` in the "Two keyword values" section to:

> The font's over edge is its start boundary, or its under edge is its end boundary, depending on which edge the value is set for. This can be considered the default, as this is where the boundaries are normally without any manipulation.

Does that sound accurate to you?

Author: @kojiishi This also gets me on to the overriding question about this PR — how do the `text-box-*` values affect the leading/half-leading? Is the half-leading stripped from the top and bottom of the text box only? Is it not affected at all? Is it only affected when certain values are set? @estelle and I have been confused by this throughout this work, as evidenced by comments like these:

- https://github.com/mdn/content/pull/37738#discussion_r1941125889
- https://github.com/mdn/content/pull/37738#discussion_r1941128995
- https://github.com/mdn/content/pull/37738#discussion_r1940314606

Your comment above implies that it isn't affected at all...but we are not sure. Can you clear up this confusion for us?

Thanks 😄

Reviewer: Yeah, this part is a bit complicated and confusing. I replied to the first comment above, but technically speaking, the trimming is applied to the block container, not to inline box nor to text box, using the metrics of the font of the block container. [This jsbin](https://jsbin.com/pederax/edit?html,output) shows the block container is trimmed, but there are no changes to inline boxes. You can also see in the example, when trimming the block container, only the font of the block container is used, ignoring descendant inline boxes such as `<big>`.

When `text-box-trim` is enabled to trim edges , it computes how much it trims. The half-leading is always excluded from the trimming point (i.e., no values of `text-box-edge` include the half-leading). Other `text-box-edge` values than `text` allow authors to move the trimming point further inside the font metrics.

Another property currently under early discussions, `line-fit-edge` is supposed to do similar trimming to inline boxes and text boxes. I guess that may make reading the spec a bit more complicated, sorry about that.

Author: OK, so the trimming is applied to the block-start and/or block-end edge of the text's block container. That makes sense.

Then we come to the effects of the different value. So, if I understand it correctly, setting `text-box-trim` to non-`none` values serves to trim the half-leading from the block container, even if no `text-box-edge` value is set? And it also follows that `text-box-edge: text` will trim the half-leading, and values like `cap`/`ex`/`alphabetic` will trim the leading plus some of the inner font space?

 I've done a bit of testing, and this seems to make sense, although you don't seem to see an effect with just `text-box-trim` in the examples I've written. To see an effect, you need to include a child element with a bigger `font-size` set, like in your example?

This explanation has already really helped my understanding, thank you! I will start to make some updates while I wait for clarification/confirmation.

Author: Update: My latest commit aims to make the text more accurate, based on @kojiishi's explanation.

Author: I'm sorry, @kojiishi. Can you please confirm that my understanding of what these properties are doing is correct, as written in the above comment?

Also, can have a quick look at the `text-edge` reference page and check whether I've got the value descriptions correct: https://pr37738.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/text-edge?

Many thanks.

Reviewer: > So, if I understand it correctly, setting text-box-trim to non-none values serves to trim the half-leading from the block container, even if no text-box-edge value is set?

Yes, because the `text-box-edge` is set to its initial value, `text`. You can consider `text-box-trim` controls on/off of the trimming, and the trimming point is set to exclude half-leading by default.

> And it also follows that text-box-edge: text will trim the half-leading, and values like cap/ex/alphabetic will trim the leading plus some of the inner font space?

Yes.

> Is the half-leading within the line-box or the text-box? Can you clarify this too?

As far as I understand, the term "text-box" isn't clearly defined in the CSS spec. I assume you mean a layout-time object that corresponds to a `Text` node? The bounding of such box isn't clearly defined, as `getClientRects()` is only for `Element`, and that geometry of a `Text` node isn't measurable.

The [CSS 2 spec 10.8.1. Leading and half-leading](https://drafts.csswg.org/css2/#leading) says:
> The height of the inline box encloses all glyphs and their half-leading on each side and is thus exactly [line-height](https://drafts.csswg.org/css2/#propdef-line-height).

so it is included in an inline box.

Also note that the term "line box" is a bit ambiguous. Some people use it to refer "inline box" (such as `<span>`), some to refer "root line box" (the object that represents a line). I'm not sure which you meant, but both includes the half-leading.

Reviewer: [Another jsbin example](https://jsbin.com/yinupem/edit?html,output) that may help you to understand how this property works. As you can see in this example, this property determines the trimming point not from the actual line, but only from a virtual line that has text of the font of the block container (i.e., a line produced by `<div>text</div>`). I'll defer what's the best way to explain this to you as you're the expert of that part.

Author: Thanks for all the further clarifications, @kojiishi! I've made a few adjustments, which will be in my next commit, but I'm fairly happy with the text now. I'm unsure how much it is worth digging into exactly where the half-leading is included, as it could start to make the docs confusing for many readers. We could maybe say some more about it on the leading glossary page. 

What do you think @estelle? I think we are just about ready to wrap this one up now.

---
title: <text-edge>
slug: Web/CSS/text-edge
page-type: css-type
browser-compat: css.types.text-edge
---

{{CSSRef}}

> The **`<text-edge>`** {{glossary("enumerated")}} [data type](/en-US/docs/Web/CSS/CSS_Types) defines keywords that specify font metrics representing specific regions on a font's block-start edge and block-end edge that are to be manipulated in some way. It effectively does this by specifying a position for the font's over and/or under edge.

Reviewer 2: ```suggestion
The **`<text-edge>`** {{glossary("enumerated")}} [data type](/en-US/docs/Web/CSS/CSS_Types) defines keywords that specify font metrics representing specific regions on a font's block-start edge and block-end edge. Each keyword specifies a position of a font's over and/or under edge.
```
data type pages are written in a property-agnostic way, even if used by a single property.

> The `text-edge` values are used in the {{cssxref("text-box-edge")}} property to define regions of spacing to trim from the block-start edge and block-end edge of specified text content.

Reviewer 2: ```suggestion
The `<text-edge>` key terms are values of the {{cssxref("text-box-edge")}} property used to specify trim definitions from the block-start edge and block-end edge of text content.
```

Author: I didn't like this edit very much. It is technically correct, but I think it will be significantly harder for readers to understand. I have elected to echo the definition from the `text-box-edge` page, and keep it really simple:

> The `<text-edge>` values are used in the {{cssxref("text-box-edge")}} property to specify an amount of space to trim from text content.

- When two values are specified, the first value specifies the position of the font's over edge, and the second value specifies the position of the font's under edge
- When only one value is specified, both edges are assigned that same keyword.

### Two keyword values

- `alphabetic`
  - : The font's under edge is its alphabetic baseline, which is the bottom of its short lower-case letters (for example, "m", "n", and "o") or capital letters.
- `cap`
  - : The font's over edge is its cap-height baseline, which is the top of its capital letters.
- `ex`
  - : The font's over edge is its x-height baseline, which is the top of its short lower-case letters.
- `text`
  - : The font's over or under edge is its start and end boundaries. This can be considered the default, as this is where the boundaries are normally without any manipulation.

### Single keyword values

- `text`
  - : The font's over and under edges are its start and end boundaries. This can be considered the default, as this is where the boundaries are normally without any manipulation.

> [!NOTE]

> > The specification also defines the `ideographic` and `ideographic-ink` keywords, which are intended to represent over and under edge positions specific to [CJK language characters](https://en.wikipedia.org/wiki/CJK_characters). Currently their exact behavior is being debated and they are are not supported by any browser.

Reviewer 2: optional: I would put single (and the note) above two keyword values.

Author: Yeah, I agree with you — that makes more logical sense for the flow. I've swapped the order over, and I've also swapped the order of the two bullets that precede those sections so that the one > two order is consistent everywhere.

## Examples

> 
> ### Valid text-edge values

Reviewer 2: ```suggestion
```

### Single keyword values

- `text`

>   - : The font's over and under edges are its start and end boundaries, which include the font's ascenders and decenders but exclude the half-leading set on the text. This can be considered the default, as this is where the boundaries are normally without any manipulation.

Reviewer 2: My understanding is that half-leading is set by `line-height`, and not a font feature.

Reviewer: True, but I think this text is good, clarifying what's included and what's not in the "start and end boundaries" is very helpful. If you worry about "exclude" may indicate that it's included in font features, maybe "without including" is more accurate?

Author: I think the current text is OK in that regard, but I have added a note to say that the half-leading is controlled by `line-height`, just so that is clear.
